pkg/memstore/memstore.go

```diff
@@ -0,0 +1,63 @@
+package memstore
+
+import (
+	"encoding/json"
+	"sync"
+)
+
+type Bucket interface {
+	Set(string, interface{})
+	Get(string) interface{}
+	MarshallJSON() ([]byte, error)
+}
+
+type Store interface {
+	Bucket(string) Bucket
+}
+
+type store struct {
+	m       *sync.Mutex
+	buckets map[string]*bucket
+}
+
+type bucket struct {
+	m      *sync.RWMutex
+	values map[string]interface{}
+}
+
+func New() Store {
+	return &store{
+		m:       &sync.Mutex{},
+		buckets: map[string]*bucket{},
+	}
+}
+
+func (s store) Bucket(name string) Bucket {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	if _, ok := s.buckets[name]; !ok {
+		s.buckets[name] = &bucket{
+			m:      &sync.RWMutex{},
+			values: map[string]interface{}{},
+		}
+	}
+
+	return s.buckets[name]
+}
+
+func (b bucket) Set(key string, value interface{}) {
+	b.m.Lock()
+	defer b.m.Unlock()
+	b.values[key] = value
+}
+
+func (b bucket) Get(key string) interface{} {
+	b.m.RLock()
```

Comment:
I don't think we need a read lock

Reply:
There's a race condition here that'll panic if we don't lock, I think that's the proper usage of read/write mutex, isn't it ?

Reply:
Ok I wasn't aware of possible concurrency even from just reading from a map. In my mind map operation are atomic and thus you do not have to lock for a read. Good to know

Reply:
I wasn't sure too so I launched tests with `-race` and it caused panic